libappstream-builder/plugins/asb-plugin-sugar-activity.c

```diff
@@ -0,0 +1,286 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+#include <string.h>
+#include <fnmatch.h>
+#include <gdk-pixbuf/gdk-pixbuf.h>
+
+#include <asb-plugin.h>
+
+#define __APPSTREAM_GLIB_PRIVATE_H
+#include <as-utils-private.h>
+#include <as-app-private.h>
+
+/**
+ * asb_plugin_get_name:
+ */
+const gchar *
+asb_plugin_get_name (void)
+{
+	return "sugar-activity";
+}
+
+#define _SUGAR_ACTIVITY_GLOB "/usr/share/sugar/activities/*/activity/activity.info"
+#define _SUGAR_ACTIVITY_ICON_GLOB "/usr/share/sugar/activities/*/activity/*"
+#define _SUGAR_ACTIVITY_LINFO_GLOB "/usr/share/sugar/activities/*/locale/*/activity.linfo"
+
+/**
+ * asb_plugin_add_globs:
+ */
+void
+asb_plugin_add_globs (AsbPlugin *plugin, GPtrArray *globs)
+{
+	// the _SUGAR_ACTIVITY_GLOB is a subset of the icon glob
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_ICON_GLOB);
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_LINFO_GLOB);
+}
+
+/**
+ * _asb_plugin_is_actinfo
+ */
+static gboolean
+_asb_plugin_is_actinfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * _asb_plugin_is_linfo
+ */
+static gboolean
+_asb_plugin_is_linfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_LINFO_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * asb_plugin_check_filename:
+ */
+gboolean
+asb_plugin_check_filename (AsbPlugin *plugin, const gchar *filename)
+{
+	return _asb_plugin_is_linfo (filename) ||
+	       _asb_plugin_is_actinfo (filename);
+}
+
+static gboolean
+_asb_plugin_load_sugar_icon (const gchar *path,
+			     AsIcon *ic) {
+	gsize len;
+	gchar *data = NULL;
+	GBytes *bytes = NULL;
+
+	if (!g_file_get_contents (path, &data, &len, NULL))
+		return FALSE;
+	printf ("ICON (got contents) = %s\n", path);
+
+	as_icon_set_kind (ic, AS_ICON_KIND_EMBEDDED);
+	bytes = g_bytes_new_take (data, len);
+	as_icon_set_data (ic, bytes);
+	as_icon_set_width (ic, 75);
+	as_icon_set_height (ic, 75);
+	return TRUE;
+}
+
+/**
+ * asb_plugin_process_actinfo:
+ */
+static gboolean
+asb_plugin_process_actinfo (AsbPlugin *plugin,
+			    AsbPackage *pkg,
+			    const gchar *filename,
+			    AsApp *app,
+			    GError **error)
+{
+	gsize len;
+	const gchar *tmp;
+	g_autofree gchar *data = NULL;
+	g_autofree gchar *id = NULL;
+	g_autoptr(GKeyFile) actinfo = NULL;
+
+	if (!g_file_get_contents (filename, &data, &len, error))
+		return FALSE;
+
+	actinfo = g_key_file_new ();
+	if (!g_key_file_load_from_data (actinfo, data, len,
+					G_KEY_FILE_NONE, error))
+		return FALSE;
+
+	if (!g_key_file_has_group (actinfo, "Activity"))
+		return FALSE;
+	if (!g_key_file_has_key (actinfo, "Activity", "bundle_id", error))
+		return FALSE;
+
+	as_app_add_category (AS_APP (app), "Sugar");
+	as_app_add_category (AS_APP (app), "SugarActivity");
```

Comment:
No.  Is there another way that we can mark sugar activities as being sugar activities in the appstream data?